### PR TITLE
Storage Transfer Manager: Job Engine Implementation

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/job_engine.hpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+      _internal::MovablePtr<JobPlan> m_jobPlan;
+      uint32_t m_id = 0;
+      size_t m_numDoneBits = 0;
+      std::unique_ptr<std::atomic<size_t>> m_numUndoneBits
+          = std::make_unique<std::atomic<size_t>>(0);
+      std::unique_ptr<_internal::MemoryMap> m_mappedFile;
+      bool* m_doneBitmap = nullptr;
+    };
+
+    struct JobPlan final
```
Comment (EmmaZhu) [inline sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/job_engine.hpp:128]: Add some comments to brief its functionality and logic flow.

File: sdk/storage/azure-storage-datamovement/src/job_engine.cpp [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+        _internal::HydrationParameters hydrateParameters)
+    {
+      // TODO: exception handling during job creation/resumption
+      auto uuid = Core::Uuid::CreateUuid().ToString();
+      EngineOperation op;
+      op.JobId = uuid;
+      op.Type = decltype(op.Type)::CreateJob;
+      op.Model = model;
+
+      EngineOperation op2;
```
Comment (EmmaZhu) [inline sdk/storage/azure-storage-datamovement/src/job_engine.cpp:204]: Why need both of create and resume here?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-datamovement/src/job_engine.cpp:204]: _Create_ operation only creates job plan files on disk. It doesn't start the job. A job after creation is in a status as if it was paused. So we need to resume it.

File: sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/job_engine.hpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/job_engine.hpp:154]: Added

File: sdk/storage/azure-storage-datamovement/src/job_engine_serialization.cpp [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+      AZURE_ASSERT(!_internal::PathExists(jobPlanDir));
+      _internal::CreateDirectory(jobPlanDir);
+      {
+        const std::string partGensFilename = _internal::JoinPath(jobPlanDir, "part_gens");
+        std::fstream fOut(
+            partGensFilename + ".tmp",
+            std::fstream::out | std::fstream::trunc | std::fstream::binary);
+        fOut.exceptions(std::fstream::failbit | std::fstream::badbit);
+        WriteFixedInt(fOut, g_PlanFileVersion);
+        PartGenerator rootGenerator;
```
Comment (EmmaZhu) [inline sdk/storage/azure-storage-datamovement/src/job_engine_serialization.cpp]: rootGenerator is not assigned any value, just serialized its default value to disk?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-datamovement/src/job_engine_serialization.cpp]: Yes, empty means this is root generator.

**Review by EmmaZhu — APPROVED** [review]

**Pull request merged**